Storage/storage.h

```diff
@@ -73,22 +73,47 @@ namespace storage {
+    using T_KEY = current::decay<decltype(::current::storage::sfinae::GetKey(std::declval<base>()))>; \
+    T_KEY key() const { return ::current::storage::sfinae::GetKey(data); }                            \
+    void set_key(const T_KEY& key) { ::current::storage::sfinae::SetKey(data, key); }                 \
+  };                                                                                                  \
+  CURRENT_STRUCT(alias##Deleted) {                                                                    \
+    CURRENT_FIELD(key, ::current::storage::sfinae::ENTRY_KEY_TYPE<base>);                             \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##Deleted) {}                                                    \
+    CURRENT_CONSTRUCTOR(alias##Deleted)(const base& value)                                            \
+        : key(::current::storage::sfinae::GetKey(value)) {}                                           \
+  };                                                                                                  \
+  struct alias {                                                                                      \
+    using T_ENTRY = base;                                                                             \
+    using T_UPDATE_EVENT = alias##Updated;                                                            \
+    using T_DELETE_EVENT = alias##Deleted;                                                            \
+    using T_PERSISTED_EVENT_1 = alias##Updated;                                                       \
+    using T_PERSISTED_EVENT_2 = alias##Deleted;                                                       \
   }
+
+#define CURRENT_STRUCT_TAG_Vector(base, alias)                                    \
+  CURRENT_STRUCT(alias##PushedElement) {                                          \
+    CURRENT_FIELD(data, base);                                                    \
+    CURRENT_DEFAULT_CONSTRUCTOR(alias##PushedElement) {}                          \
+    CURRENT_CONSTRUCTOR(alias##PushedElement)(const base& value) : data(value) {} \
+  };                                                                              \
+  CURRENT_STRUCT(alias##PoppedElement) {};                                        \
+  struct alias {                                                                  \
+    using T_ENTRY = base;                                                         \
+    using T_PUSHED_BACK_EVENT = alias##PushedElement;                             \
+    using T_POPPED_BACK_EVENT = alias##PoppedElement;                             \
+    using T_PERSISTED_EVENT_1 = alias##PushedElement;                             \
+    using T_PERSISTED_EVENT_2 = alias##PoppedElement;                             \
+  }
+
+// TODO(dkorolev) || TODO(mzhurovich): Move to `TypeSystem/` and test.
+#define CURRENT_STRUCT_TAG(tag, base, alias) CURRENT_STRUCT_TAG_##tag(base, alias)
```

Comment:
Yes, tagging is a way to tag any `CURRENT_STRUCT`, with possibly a custom behavior.

How about we flip the parameters and allow two "signatures": `CURRENT_STRUCT_TAG(original_struct_name, tagged_struct_name)` and `CURRENT_STRUCT_TAG(original_struct_name, tagged_struct_name, tag_type)`?

As long as tagged names can be used later, and (for some tags, at least) there is a way to retrieve the "base" structure from its tagged type, we should be good to go.

Practical example: `CURRENT_STRUCT_TAG(User, HypermediaUserResponse, HypermediaGETResponse);`